sim/paladin/paladin.go

```diff
@@ -0,0 +1,175 @@
+	// HammerOfWrath         *core.Spell
+	// HolyWrath             *core.Spell
+
+	// Seal spells and their associated auras
+	SealOfRighteousness []*core.Spell
+	SealOfCommand       []*core.Spell
+	SealOfMartyrdom     *core.Spell
+
+	SealOfRighteousnessAura []*core.Aura
+	SealOfCommandAura       []*core.Aura
+	SealOfMartyrdomAura     *core.Aura
+
+	// Auras from talents
+	DivineFavorAura *core.Aura
+	VengeanceAura   *core.Aura
+
+	// Placeholder for any auto-rotation with exo/HW.
+	DemonAndUndeadTargetCount int32
+}
+
+// Implemented by each Paladin spec.
+type PaladinAgent interface {
+	GetPaladin() *Paladin
+}
+
+func (paladin *Paladin) GetCharacter() *core.Character {
+	return &paladin.Character
+}
+
+func (paladin *Paladin) GetPaladin() *Paladin {
+	return paladin
+}
+
+func (paladin *Paladin) AddRaidBuffs(raidBuffs *proto.RaidBuffs) {
+	// Buffs are handled explicitly through APLs now
+}
+
+func (paladin *Paladin) AddPartyBuffs(_ *proto.PartyBuffs) {
+}
+
+func (paladin *Paladin) Initialize() {
+	// Update auto crit multipliers now that we have the targets.
+	paladin.AutoAttacks.MHConfig().CritMultiplier = paladin.MeleeCritMultiplier()
+
+	// Judgement and Seals
+	paladin.registerJudgementSpell()
+	paladin.registerSealOfRighteousnessSpellAndAura()
+
+	paladin.registerSealOfCommandSpellAndAura()
+	paladin.registerSealOfMartyrdomSpellAndAura()
+
+	// Active abilities
+	paladin.registerCrusaderStrikeSpell()
+	paladin.registerDivineStormSpell()
+	paladin.registerConsecrationSpell()
+	paladin.registerHolyShockSpell()
+	paladin.registerExorcismSpell()
+	paladin.registerDivineFavorSpellAndAura()
+	// paladin.registerHammerOfWrathSpell()
+	// paladin.registerHolyWrathSpell()
+	// paladin.registerHolyShieldSpell()
+
+	for i := int32(0); i < paladin.Env.GetNumTargets(); i++ {
+		unit := paladin.Env.GetTargetUnit(i)
+		if unit.MobType == proto.MobType_MobTypeDemon || unit.MobType == proto.MobType_MobTypeUndead {
+			paladin.DemonAndUndeadTargetCount += 1
+		}
+	}
+}
+
+func (paladin *Paladin) Reset(_ *core.Simulation) {
+	paladin.CurrentSeal = nil
+}
+
+// maybe need to add stat dependencies
+func NewPaladin(character *core.Character, talentsStr string) *Paladin {
+	paladin := &Paladin{
+		Character: *character,
+		Talents:   &proto.PaladinTalents{},
+	}
+	core.FillTalentsProto(paladin.Talents.ProtoReflect(), talentsStr, TalentTreeSizes)
+
+	paladin.PseudoStats.CanParry = true
+	paladin.EnableManaBar()
+	paladin.AddStatDependency(stats.Strength, stats.AttackPower, 2.0)
+	paladin.AddStatDependency(stats.Agility, stats.MeleeCrit, core.CritPerAgiAtLevel[character.Class][int(paladin.Level)]*core.CritRatingPerCritChance)
+	paladin.AddStatDependency(stats.Intellect, stats.SpellCrit, core.CritPerIntAtLevel[character.Class][int(paladin.Level)]*core.SpellCritRatingPerCritChance)
+
+	// Paladins get 0.0167 dodge per agi. ~1% per 59.88
+	paladin.AddStatDependency(stats.Agility, stats.Dodge, (1.0/59.88)*core.DodgeRatingPerDodgeChance)
```

Comment:
I'm informed that it does scale with level in classic. 

e just generally seem to be missing a proper `CritPerAgiAtLevel` table analogue for dodge in the SoD code at present, no other specs, not even those with tank specs enabled, implement this at present.

Also, if this data is trustworthy it would imply the BaseDodge is wrong a few lines later:
https://www.wowhead.com/classic/guide/classic-wow-stats-and-attributes-overview#dodge

I can get someone to test this in-game.

Reply:
I think this link is only for level 60. You can find stats from the RatingBuster addong that VoraciousGhost(in the sim discord) maintains. We have used his data for the crit ratings formulas

Reply:
https://github.com/raethkcj/RatingBuster/blob/master/libs/StatLogic/Vanilla_Logic.lua#L121

Relevant line, appears to be the same as `CritPerAgiAtLevel` for paladins. It seems like this should have its own function to handle special cases for rogues and be human readable.

I could change this to
`	paladin.AddStatDependency(stats.Agility, stats.Dodge, core.CritPerAgiAtLevel[character.Class][int(paladin.Level)]*core.DodgeRatingPerDodgeChance)`

or, given we're not implementing the prot sim atm where avoidance stats are actually relevant, we could just comment out the stuff related to basedodge/baseparry and the dodge agi dependency with relevant comments; then we'd add this when there's an appropriately named function in core like `DodgePerAgiAtLevel` to prevent confusion.

What do you think?

Reply:
you can comment it out for now if its not relevant yes

Reply:
Addressed in https://github.com/wowsims/sod/pull/302/commits/0cd88fb8da4ba855a85b24ec39f2aab83d28f550